File: fortran/empsvd_core.f90
module empsvd_core
  implicit none
  
  ! mandatory variables to be initialize by init subroutine
  real(8)   , protected, allocatable :: x(:), y(:)
  real(8)   , protected, allocatable :: theta(:, :)
  real(8)   , protected, allocatable :: gamm(:, :)
  integer(8), protected              :: N, K
  integer(8), public   , parameter   :: M = 6
  integer(8), protected              :: niter = 0

  real(8), private :: tol_ = 1d-2
  integer(8), private :: max_iter_ = 1000
  logical, private :: fix_ab_ = .false.
  logical, private :: fix_alpha_ = .true.

  ! mandatory subroutine
  public :: init, fit, e_step, m_step
  ! optional subroutine
  public :: get_loglikelihood, get_aic, get_bic

  private :: calc_new_pk, calc_new_ak, calc_new_bk, calc_new_sk, calc_new_lk
  private :: calc_log_pxy, calc_pxy, calc_logsum_pxy, calc_sum_pxy, calc_loglikelihood


contains


  subroutine init(x_in, y_in, k_in, theta0, max_iter, tol, fix_alpha, fix_ab)
    use empsvd_static, only: stop_with_error, make_theta0
    implicit none
    real(8)   , intent(in) :: x_in(:), y_in(:)
    integer(8), intent(in) :: k_in
    real(8)   , intent(in), optional :: theta0(:, :)
    integer(8), intent(in), optional :: max_iter
    real(8)   , intent(in), optional :: tol
    logical   , intent(in), optional :: fix_alpha, fix_ab
    integer(8) :: shp(2)

    if ( size(x_in) /= size(y_in) ) call stop_with_error("data length of x and y must be same.")
    if ( present(theta0) ) then
       shp = shape(theta0)
       if ( shp(1) /= k_in ) call stop_with_error("Invalid shape of theta0.")
       if ( shp(2) /= M ) call stop_with_error("Invalid shape of theta0.")
    end if
    
    if ( allocated(x) ) deallocate(x)
    if ( allocated(y) ) deallocate(y)
    if ( allocated(theta) ) deallocate(theta)
    if ( allocated(gamm) ) deallocate(gamm)

    N = size(x_in)
    K = k_in

    allocate( x(N) ) 
    allocate( y(N) )
    allocate( theta(K, M) )
    allocate( gamm(N, K) )

    if ( present(max_iter) ) max_iter_ = max_iter
    if ( present(tol) ) tol_ = tol
    if ( present(fix_alpha) ) fix_alpha_ = fix_alpha
    if ( present(fix_ab) ) fix_ab_ fix_ab

    x = x_in
    y = y_in

    if ( present(theta0) ) then
       theta = theta0
    else
       call make_theta0(theta)
    end if
       
    call calc_gamma(gamm)

    niter = 0

  end subroutine init


  subroutine fit(info)
    implicit none
    integer(8), intent(out) :: info
    real(8) :: l1, l2

    call get_loglikelihood(l1)
    do

       if ( niter > max_iter_ ) then
          info = 1
          return
       end if

       call e_step()
       call m_step(info)
       if (info > 0) return

       call get_loglikelihood(l2)

       if ( abs(l1 - l2) < tol_ ) then
          info = 0
          exit
       end if

       l1 = l2
       niter = niter + 1

    end do
    
  end subroutine fit


  subroutine get_bic(bic)
    implicit none
    real(8), intent(out) :: bic
    real(8) :: nn, kk, mm

    nn = real(N)
    kk = real(K)
    mm = real(M)
    
    call get_loglikelihood(bic)
    bic = bic - 0.5d0 * kk * mm* log(nn)

  end subroutine get_bic


  subroutine get_aic(aic)
    implicit none
    real(8), intent(out) :: aic
    real(8) :: kk, mm

    kk = real(K)
    mm = real(M)

    call get_loglikelihood(aic)
    aic = aic - kk * mm

  end subroutine get_aic


  subroutine get_loglikelihood(loglikelihood)
    implicit none
    real(8), intent(out) :: loglikelihood
    
    call calc_loglikelihood(loglikelihood)

  end subroutine get_loglikelihood


  subroutine calc_log_pxy(log_pxy, theta_in)
    use empsvd_static, only: calc_log_pxy_ => calc_log_pxy
    implicit none
    real(8), intent(out)           :: log_pxy(N, K)
    real(8), intent(in) , optional :: theta_in(K, M)

    if ( present(theta_in) ) then
       call calc_log_pxy_(x, y, theta_in, log_pxy)
    else
       call calc_log_pxy_(x, y, theta, log_pxy)
    end if
    
  end subroutine calc_log_pxy


  subroutine calc_pxy(pxy, theta_in)
    use empsvd_static, only: calc_pxy_ => calc_pxy
    implicit none
    real(8), intent(out) :: pxy(N, K)
    real(8), intent(in), optional :: theta_in(K, M)

    if ( present(theta_in) ) then
       call calc_pxy_(x, y, theta_in, pxy)
    else
       call calc_pxy_(x, y, theta, pxy)
    end if

  end subroutine calc_pxy


  subroutine calc_logsum_pxy(logsum_pxy, theta_in)
    implicit none
    real(8), intent(out) :: logsum_pxy(N)
    real(8), intent(in) , optional :: theta_in(K, M)
    real(8) :: log_pxy(N, K), dlog_pxy(N, K), max_log_pxy(N)
    integer(8) :: j

    if ( present(theta_in) ) then
       call calc_log_pxy(log_pxy, theta_in)
    else
       call calc_log_pxy(log_pxy)
    end if

    max_log_pxy = maxval( log_pxy, dim=2 )
    do j = 1, K
       dlog_pxy(:, j) = log_pxy(:, j) - max_log_pxy
    end do
    logsum_pxy = log( sum( exp(dlog_pxy), dim=2 ) ) + max_log_pxy

  end subroutine calc_logsum_pxy

  
  subroutine calc_sum_pxy(sum_pxy, theta_in)
    implicit none
    real(8), intent(out) :: sum_pxy(N)
    real(8), intent(in) , optional :: theta_in(K, M)
    
    if ( present(theta_in) ) then
       call calc_logsum_pxy(sum_pxy, theta_in)
    else
       call calc_logsum_pxy(sum_pxy)
    end if
    sum_pxy = exp(sum_pxy)

  end subroutine calc_sum_pxy


  subroutine calc_loglikelihood(loglikelihood, theta_in)
    implicit none
    real(8), intent(out) :: loglikelihood
    real(8), intent(in) , optional :: theta_in(K, M)
    real(8) :: logsum_pxy(N)

    if ( present(theta_in) ) then
       call calc_logsum_pxy(logsum_pxy, theta_in)
    else
       call calc_logsum_pxy(logsum_pxy)
    end if

    loglikelihood = sum( logsum_pxy )

  end subroutine calc_loglikelihood


  subroutine calc_gamma(g)
    implicit none
    real(8), intent(out) :: g(N, K)
    real(8) :: sum_pxy(N)
    integer(8) :: j

    call calc_sum_pxy(sum_pxy)
    call calc_pxy(g)
    
    do j = 1, K
       g(:, j) = g(:, j) / sum_pxy
    end do
    
  end subroutine calc_gamma


  subroutine e_step()
    implicit none

    call calc_gamma(gamm)

  end subroutine e_step


  subroutine m_step(info)
    implicit none
    integer(8), intent(out) :: info
    integer(8) :: j
    real(8)    :: tmp

    do j = 1, K

       if ( .not. fix_ab_ ) then
          call calc_new_bk(gamm(:, j), theta(j, 3), tmp, info)
          if ( info > 0 ) return
          theta(j, 3) = tmp
          call calc_new_ak(gamm(:, j), theta(j, 3), theta(j, 2))
       end if

       call calc_new_sk(gamm(:, j), theta(j, 2), theta(j, 3), theta(j, 4))

       if ( .not. fix_alpha_ ) then
          call calc_new_alk(gamm(:, j), theta(j, 5), tmp, info)
          if ( info > 0 ) return
          theta(j, 5) = tmp
       end if

       call calc_new_lk(gamm(:, j), theta(j, 5), theta(j, 6))
       call calc_new_pk(gamm(:, j), theta(j, 1))
       
    end do

    info = 0
    
  end subroutine m_step


  subroutine calc_new_pk(gammak, new_pk)
    implicit none
    real(8), intent(in)  :: gammak(N)
    real(8), intent(out) :: new_pk

    new_pk = sum( gammak ) / real(N)

  end subroutine calc_new_pk


  subroutine calc_new_ak(gammak, bk, new_ak)
    use empsvd_static, only: calc_new_ak_ => calc_new_ak
    implicit none
    real(8), intent(in)  :: gammak(N)
    real(8), intent(in)  :: bk
    real(8), intent(out) :: new_ak

    call calc_new_ak_(bk, x, y, gammak, new_ak)
    
  end subroutine calc_new_ak


  subroutine calc_new_bk(gammak, old_bk, new_bk, info)
    use empsvd_static, only: calc_new_bk_by_newton, get_rand
    implicit none
    real(8), intent(in)  :: gammak(N), old_bk
    real(8), intent(out) :: new_bk
    integer(8), intent(out) :: info
    integer(8), parameter :: retry = 30
    real(8)    :: bk1(retry)
    integer(8) :: i

    call calc_new_bk_by_newton(old_bk, x, y, gammak, new_bk, info)
    if ( info /= 0) then

       call get_rand(bk1)
       do i = 1, retry
!          write(*, "(A,f0.2)") "new_bk: retry from another bk1 ", old_bk + bk1(i)
          call calc_new_bk_by_newton(old_bk, x, y, gammak, new_bk, info, bk1(i))

          if ( info == 0 ) return
       end do
    end if

  end subroutine calc_new_bk


  subroutine calc_new_sk(gammak, ak, bk, new_sk)
    implicit none
    real(8), intent(in)  :: gammak(N)
    real(8), intent(in)  :: ak, bk
    real(8), intent(out) :: new_sk
    real(8) :: q1, q2

    q1 = sum( gammak * (y - (ak * x ** bk)) ** 2 )
    q2 = sum( gammak )
    new_sk = q1 / q2
    
  end subroutine calc_new_sk


  subroutine calc_new_alk(gammak, old_alk, new_alk, info)
    use empsvd_static, only: calc_new_alk_by_invdigamma
    implicit none
    real(8), intent(in)  :: gammak(N)
    real(8), intent(in)  :: old_alk
    real(8), intent(out) :: new_alk
    integer(8), intent(out) :: info

    call calc_new_alk_by_invdigamma(old_alk, x, gammak, new_alk, info)

  end subroutine calc_new_alk


  subroutine calc_new_lk(gammak, alk, new_lk)
    use empsvd_static, only: calc_new_lk_ => calc_new_lk
    implicit none
    real(8), intent(in)  :: gammak(N), alk
    real(8), intent(out) :: new_lk
    
    call calc_new_lk_(alk, x, gammak, new_lk)

  end subroutine calc_new_lk


end module empsvd_core

File: cpp/EmpsvdCore.cpp
#include "EmpsvdCore.h"
#include <Eigen/Dense>
#include <cmath>
#include <stdexcept>
#include <cstddef>


const size_t Empsvd::EmpsvdCore::m;


Empsvd::EmpsvdCore::EmpsvdCore(
	size_t k, const Eigen::ArrayXd& x, const Eigen::ArrayXd& y, const Eigen::ArrayXXd& theta0,
	size_t max_iter, double tol, bool fix_alpha, bool fix_ab
) : x(x), y(y), z(Eigen::ArrayXd::Ones(x.size())), k(k), theta(theta0), 
	max_iter(max_iter), tol(tol), fix_ab(fix_ab), fix_alpha(fix_alpha), n(x.size())
{
	this->check_init();
	this->gamma = this->get_gamma();
}

Empsvd::EmpsvdCore::EmpsvdCore(
	size_t k, const Eigen::ArrayXd& x, const Eigen::ArrayXd& y,
	size_t max_iter, double tol, bool fix_alpha, bool fix_ab
) : x(x), y(y), z(Eigen::ArrayXd::Ones(x.size())), k(k), 
	max_iter(max_iter), tol(tol), fix_ab(fix_ab), fix_alpha(fix_alpha), n(x.size())
{
	this->theta = this->make_theta0(x, y, k);
	this->check_init();
	this->gamma = this->get_gamma();
}

Empsvd::EmpsvdCore::EmpsvdCore(
	size_t k, const Eigen::ArrayXd& x, const Eigen::ArrayXd& y, const Eigen::ArrayXd& z, const Eigen::ArrayXXd& theta0,
	size_t max_iter, double tol, bool fix_alpha, bool fix_ab
) : x(x), y(y), z(z), k(k), theta(theta0),
	max_iter(max_iter), tol(tol), fix_ab(fix_ab), fix_alpha(fix_alpha), n(x.size())
{
	this->check_init();
	this->gamma = this->get_gamma();
}

Empsvd::EmpsvdCore::EmpsvdCore(
	size_t k, const Eigen::ArrayXd& x, const Eigen::ArrayXd& y, const Eigen::ArrayXd& z,
	size_t max_iter, double tol, bool fix_alpha, bool fix_ab
) : x(x), y(y), z(z), k(k),
	max_iter(max_iter), tol(tol), fix_ab(fix_ab), fix_alpha(fix_alpha), n(x.size())
{
	this->theta = this->make_theta0(x, y, z, k);
	this->check_init();
	this->gamma = this->get_gamma();
}

Empsvd::EmpsvdCore::~EmpsvdCore()
{
}

void Empsvd::EmpsvdCore::fit()
{
	double l1 = this->get_loglikelihood();
	double l2;

	for (int i = 0; i < this->max_iter; i++) {
		this->e_step();
		this->m_step();
		this->niter += 1;
		l2 = this->get_loglikelihood();
		if (std::abs(l2 - l1) < this->tol) return;
		l1 = l2;
	}

	throw std::runtime_error("Failed to converge EM algorithm.");
}

void Empsvd::EmpsvdCore::e_step()
{
	this->gamma = this->get_gamma();
}

void Empsvd::EmpsvdCore::m_step()
{
	for (Eigen::Index ik = 0; ik < this->k; ik++) {
		if (!this->fix_ab) {
			this->theta(ik, 2) = this->get_new_bk(ik);
			this->theta(ik, 1) = this->get_new_ak(ik, this->theta(ik, 2));
		}
		this->theta(ik, 3) = this->get_new_sigma2k(ik, this->theta(ik, 1), this->theta(ik, 2));
		if (!this->fix_alpha) {
			this->theta(ik, 4) = this->get_new_alphak(ik);
		}
		this->theta(ik, 5) = this->get_new_lambdak(ik, this->theta(ik, 4));
		this->theta(ik, 0) = this->get_new_omegak(ik);
	}
}

double Empsvd::EmpsvdCore::get_aic()
{
	double l = this->get_loglikelihood();
	return l - (this->k * this->m);
}

double Empsvd::EmpsvdCore::get_bic()
{
	double l = this->get_loglikelihood();
	return l - (0.5 * this->k * this->m * std::log(this->n));
}

double Empsvd::EmpsvdCore::get_loglikelihood()
{
	return this->get_loglikelihood(this->theta);
}

double Empsvd::EmpsvdCore::get_loglikelihood(const Eigen::ArrayXXd& theta)
{
        return (this->get_logsum_pxy(this->get_log_pxy(theta)) * this->z).sum();
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::make_theta0(
	size_t const k, const Eigen::ArrayXd& x, const Eigen::ArrayXd& y, const Eigen::ArrayXd& z
)
{
	Eigen::ArrayXXd theta0(k, Empsvd::EmpsvdCore::m);

	double const n = z.sum();
	double const y_mean = (y * z).sum() / n;
	double const y_var = ((y - y_mean).square() * z).sum() / n;
	double const x_mean = (x * z).sum() / n;
	double const x_var = ((x - x_mean).square() * z).sum() / n;

	theta0.col(0) = 1. / k;                        // omega: mixing fraction
	theta0.col(3) = y_var;                         // sigma2: variance of velocity
	theta0.col(4) = std::pow(x_mean, 2) / x_var;   // alpha: mu + 1, shape parameter of size distribution
	theta0.col(5) = x_mean / x_var;                // lambda: slope parameter of size distribution

	Eigen::ArrayXd exist_data = (z.array() > 0.).select(
		Eigen::ArrayXd::Constant(z.size(), 1),
		Eigen::ArrayXd::Constant(z.size(), 0)
	);
	double a_upp = (y * exist_data).maxCoeff() / x_mean;
	double a_low = y_mean / 4;

	if (k > 1) {
		Eigen::ArrayXd lins = Eigen::ArrayXd::LinSpaced(k, 0., 1.);
		theta0.col(1) = (lins * std::log(a_upp / a_low) + std::log(a_low)).exp();  // a: an intercept parameter of v-d relationship
		theta0.col(2) = lins;                                                      // b: a slope parameter of v-d relationship
	}
	else {
		theta0.col(1) = std::exp(std::log(a_upp * a_low) / 2.);
		theta0.col(2) = 0.5;
	}

	return theta0;
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::make_theta0(size_t const k, const Eigen::ArrayXd& x, const Eigen::ArrayXd& y)
{
	return Empsvd::EmpsvdCore::make_theta0(
		x, y, Eigen::ArrayXd::Ones(x.size()), k
	);
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::calc_log_pxy(const Eigen::ArrayXd& x, const Eigen::ArrayXd& y, const Eigen::ArrayXXd& theta)
{
	size_t const n = x.size();
	size_t const k = theta.rows();
	Eigen::ArrayXXd log_pxy(n, k);

	for (Eigen::Index ik = 0; ik < k; ik++) {
		log_pxy.col(ik) = (
			theta(ik, 0) * std::pow(theta(ik, 5), theta(ik, 4)) * x.pow(theta(ik, 4) - 1) /
			(std::sqrt(2 * M_PI * theta(ik, 3)) * std::tgamma(theta(ik, 4)))
			).log() - (
			((y - (theta(ik, 1) * x.pow(theta(ik, 2)))).square() / (2 * theta(ik, 3))) +
				(theta(ik, 5) * x)
				);
	}

	return log_pxy;
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::calc_pxy(const Eigen::ArrayXd& x, const Eigen::ArrayXd& y, const Eigen::ArrayXXd& theta)
{
	return calc_log_pxy(x, y, theta).exp();
}

double Empsvd::EmpsvdCore::digammad(double a)
{
	double const g = 0.57721566490153286061;
	double dig;
	double a2 = std::floor(a);
	double a1 = a - a2;

	if (a1 != 0.) {
		Eigen::ArrayXd c(6);
		c << 0.64493313, -0.20203181, 0.08209433, -0.03591665, 0.01485925, -0.00472050;
		// first, compute the digamma value for 0 < a < 1
		dig = (a1 / (a1 + 1)) - g + (0.5 * std::pow(a1, 7));
		for (int i = 1; i <= c.size(); i++) {
			dig += c(i-1) * (std::pow(a1, i) - std::pow(a1, 7));
		}
		dig -= 1 / a1;
		// then, increment upto a >= 1
		if (a2 > 0.) {
			for (double i = 1.; i <= a2; i++) {
				dig += 1 / (a1 + i - 1);
			}
		}
	}
	else {
		// a1 == 0 means the digamma value is special value
		dig = -g;
		if (a2 > 1.) {
			for (double i = 1; i <= a2 - 1; i++) {
				dig += 1 / i;
			}
		}
	}
	return dig;
}

void Empsvd::EmpsvdCore::check_init()
{
	if (x.size() != n || y.size() != n || z.size() != n) throw std::length_error("Mismatch size between x, y, and z.");
	if (theta.cols() != m) throw std::length_error("Column length of theta must be 6.");
	if (z.sum() <= 0.) throw std::overflow_error("sum(z)=0: This will cause overflow.");
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::get_log_pxy()
{
	return this->get_log_pxy(this->theta);
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::get_log_pxy(const Eigen::ArrayXXd& theta)
{
	return this->calc_log_pxy(this->x, this->y, theta);
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::get_pxy()
{
	return this->get_pxy(this->theta);
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::get_pxy(const Eigen::ArrayXXd& theta)
{
	return this->calc_pxy(this->x, this->y, theta);
}

Eigen::ArrayXd Empsvd::EmpsvdCore::get_logsum_pxy()
{
	return this->get_logsum_pxy(this->get_log_pxy());
}

Eigen::ArrayXd Empsvd::EmpsvdCore::get_logsum_pxy(Eigen::ArrayXXd log_pxy)
{
	Eigen::ArrayXd max_log_pxy = log_pxy.rowwise().maxCoeff();

	for (Eigen::Index ik = 0; ik < this->k; ik++) {
		log_pxy.col(ik) -= max_log_pxy;
	}
	return log_pxy.exp().rowwise().sum().log() + max_log_pxy;
}

Eigen::ArrayXd Empsvd::EmpsvdCore::get_sum_pxy()
{
	return this->get_sum_pxy(this->get_log_pxy());
}

Eigen::ArrayXd Empsvd::EmpsvdCore::get_sum_pxy(Eigen::ArrayXXd log_pxy)
{
	return this->get_logsum_pxy(log_pxy).exp();
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::get_gamma()
{
	return this->get_gamma(this->theta);
}

Eigen::ArrayXXd Empsvd::EmpsvdCore::get_gamma(const Eigen::ArrayXXd& theta)
{
	Eigen::ArrayXXd pxy = this->get_pxy(theta);
	Eigen::ArrayXd sum_pxy = this->get_sum_pxy(pxy.log());

	for (Eigen::Index ik = 0; ik < this->k; ik++) {
		pxy.col(ik) /= sum_pxy;
	}

	return pxy;
}

double Empsvd::EmpsvdCore::get_new_omegak(Eigen::Index ik)
{
	return (this->z * this->gamma.col(ik)).sum() / this->z.sum();
}

double Empsvd::EmpsvdCore::get_new_ak(Eigen::Index ik, double new_bk)
{
	double q1 = (this->z * this->gamma.col(ik) * this->y * this->x.pow(new_bk)).sum();
	double q2 = (this->z * this->gamma.col(ik) * this->x.pow(2 * new_bk)).sum();
	return q1 / q2;
}

double Empsvd::EmpsvdCore::get_new_bk(Eigen::Index ik)
{
	try
	{
		double const offset = 0.001;
		return this->get_new_bk_by_newton(ik, this->theta(ik, 2), this->theta(ik, 2) + offset);
	}
	catch (const std::runtime_error&)
	{
		size_t const max_retry = 30;
		Eigen::ArrayXd bk1 = (Eigen::ArrayXd::Random(max_retry) + 1.0) * 0.5; // [0:1] random number
		for (Eigen::Index i = 0; i < max_retry; i++) {
			try
			{
				return this->get_new_bk_by_newton(ik, this->theta(ik, 2), bk1(i));
			}
			catch (const std::runtime_error& e)
			{
				continue;
			}
		}
		throw std::runtime_error("Failed to find an optimal new bk.");
	}
}

double Empsvd::EmpsvdCore::get_new_sigma2k(Eigen::Index ik, double new_ak, double new_bk)
{
	double q1 = (this->z * this->gamma.col(ik) * (this->y - (new_ak * this->x.pow(new_bk))).square()).sum();
	double q2 = (this->z * this->gamma.col(ik)).sum();
	return q1 / q2;
}

double Empsvd::EmpsvdCore::get_new_alphak(Eigen::Index ik)
{
	return this->get_new_alphak_by_invdigamma(ik, this->theta(ik, 4));
}

double Empsvd::EmpsvdCore::get_new_lambdak(Eigen::Index ik, double new_alphak)
{
	double q1 = (this->z * this->gamma.col(ik)).sum();
	double q2 = (this->z * this->gamma.col(ik) * x).sum();
	return new_alphak * q1 / q2;
}

double Empsvd::EmpsvdCore::bkdot(Eigen::Index ik, double bk)
{
	double ak = this->get_new_ak(ik, bk);
	return (
		this->z * this->gamma.col(ik) * this->x.log() *
		((this->y * this->x.pow(bk)) - (ak * this->x.pow(2 * bk)))
		).sum();
}

double Empsvd::EmpsvdCore::get_new_bk_by_newton(Eigen::Index ik, double bk0, double bk1, size_t max_iter, double tol)
{
	double fbk0 = this->bkdot(ik, bk0);
	double fbk1 = this->bkdot(ik, bk1);
	double fbk2, bk2;

	for (int i = 0; i < max_iter; i++) {
		bk2 = ((bk0 * fbk1) - (bk1 * fbk0)) / (fbk1 - fbk0);
		fbk2 = this->bkdot(ik, bk2);

		if (std::abs(fbk2) < tol) return bk2;

		bk0 = bk1;
		fbk0 = fbk1;
		bk1 = bk2;
		fbk1 = fbk2;
	}
	throw std::runtime_error("Failed to find optimal bk with newton method.");
}

double Empsvd::EmpsvdCore::get_new_alphak_by_invdigamma(Eigen::Index ik, double alphak0, size_t max_iter, double tol)
{
	double q1 = (this->z * this->gamma.col(ik)).sum();
	double logx_mean = (this->z * this->gamma.col(ik) * this->x.log()).sum() / q1;
	double x_mean = (this->z * this->gamma.col(ik) * this->x).sum() / q1;
	double y = std::log(x_mean) - logx_mean;

	double dig, alphak1;
	
	for (int i = 0; i < max_iter; i++) {
		dig = this->digammad(alphak0);
		alphak1 = alphak0 * (std::log(alphak0) - dig) / y;
		if (std::abs(alphak1 - alphak0) < tol) return alphak1;
		alphak0 = alphak1;
	}
	throw std::runtime_error("Failed to find an optimal alpha.");
}

